documentation/examples/custom-sd/adapter-usage/main.go

```diff
@@ -0,0 +1,291 @@
+
+		tgroup := targetgroup.Group{
+			Source:  srv.name,
+			Labels:  srv.labels,
+			Targets: make([]model.LabelSet, 0, len(nodes)),
+		}
+
+		for i, node := range nodes {
+			fmt.Println("node: ", node, " i: ", i)
+			// We surround the separated list with the separator as well. This way regular expressions
+			// in relabeling rules don't have to consider tag positions.
+			var tags = srv.tagSeparator + strings.Join(node.ServiceTags, srv.tagSeparator) + srv.tagSeparator
+
+			// If the service address is not empty it should be used instead of the node address
+			// since the service may be registered remotely through a different node
+			var addr string
+			if node.ServiceAddress != "" {
+				addr = net.JoinHostPort(node.ServiceAddress, fmt.Sprintf("%d", node.ServicePort))
+			} else {
+				addr = net.JoinHostPort(node.Address, fmt.Sprintf("%d", node.ServicePort))
+			}
+			// fmt.Println("addr: ", addr)
+			target := model.LabelSet{model.AddressLabel: model.LabelValue(addr)}
+			tgroup.Labels[addressLabel] = model.LabelValue(node.Address)
+			tgroup.Labels[nodeLabel] = model.LabelValue(node.Node)
+			tgroup.Labels[tagsLabel] = model.LabelValue(tags)
+			tgroup.Labels[serviceAddressLabel] = model.LabelValue(node.ServiceAddress)
+			tgroup.Labels[servicePortLabel] = model.LabelValue(strconv.Itoa(node.ServicePort))
+			tgroup.Labels[serviceIDLabel] = model.LabelValue(node.ServiceID)
+
+			// Add all key/value pairs from the node's metadata as their own labels
+			for k, v := range node.NodeMeta {
+				name := strutil.SanitizeLabelName(k)
+				tgroup.Labels[model.LabelName(name)] = model.LabelValue(v)
+			}
+
+			tgroup.Targets = append(tgroup.Targets, target)
+		}
+		// Check context twice to ensure we always catch cancelation.
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+		select {
+		case <-ctx.Done():
+			return
+		case ch <- []*targetgroup.Group{&tgroup}:
+		}
+	}
+}
+
+// Note: This is the struct with your implementation of the Discoverer interface (see Run function).
+// Discovery retrieves target information from a Consul server and updates them via watches.
+type discovery struct {
+	client           *consul.Client
+	clientConf       *consul.Config
+	clientDatacenter string
+	tagSeparator     string
+	watchedServices  []string // Set of services which will be discovered.
+	logger           log.Logger
+}
+
+// Note: you must implement this function for your discovery implementation as part of the
+// Discoverer interface. Here you should query your SD for it's list of known targets, determine
+// which of those targets you care about (for example, which of Consuls known services do you want
+// to scrape for metrics), and then send those targets as a target.TargetGroup to the ch channel.
+func (d *discovery) Run(ctx context.Context, ch chan<- []*targetgroup.Group) {
+	// Watched services and their cancelation functions.
+	services := map[string]func(){}
+
+	var lastIndex uint64
+	for {
+		catalog := d.client.Catalog()
```

Comment:
To keep things as simple as possible I'd not use the client library or watches, and instead use the Go http client directly with a simple refresh every say 30s. Most SDs seem to be of the refresh type rather than the watch type, and if someone needs and example of watching they can always look at other SD implementations.